### ISSUE TITLE
RFC: F-API to complicated to use, P-API to limited

### ISSUE THREAD

**Issue by RoadRunnr** [Bluehouse-Technology/otter#12]
**Brainstorming Mode**

## functional API

The functional API requires very intrusive changes to existing code. It's almost like a new state variable needs to be added to very interface and passed along to other calls and the again returned to the caller (if something down the call chain is supposed to add something to the span).

For anything that has already a process state, this adds a function call (span) state. This might be ok if there already is a per function call state, but most of time there is no such thing or it's using a data type (e.g. a list) that is suitable for simply adding a new field.

Using the functional API for anything but simples call flow becomes a nightmare.

My idea/suggestion: abandon the function API

## process dictionary API

This API is too limited. There can only one span per process. It is IMHO highly desirable to be able to have sub-spans within the same process context.

# Idea

## SpanId API

This API replace the existing P-API and can be compatible with the F-API.

The basic idea is not pass along the full span state, but instead generate a SpanId when start() is called, pass that into the other function and don't change it.

API:

```
%% ====================  SPAN API  ======================
-spec start(Info :: info()) -> span_id().
-spec start(Info :: info(), integer()) -> span_id().
-spec start(Info :: info(), integer(), integer()) -> span_id().

-spec tag(SpanId :: span_id(), Key :: info(), Value :: info()) -> span_id().
-spec tag(SpanId :: span_id(), Key :: info(), Value :: info(), Service :: service()) -> span_id().

-spec log(SpanId :: span_id(), LogMsg :: info()) -> span_id().
-spec log(SpanId :: span_id(), LogMsg :: info(), service()) -> span_id().

-spec finish(SpanId :: span_id()) -> ok.

-spec ids(SpanId :: span_id()) -> {trace_span_id(), span_id()}.
```

* the functions would return the SpanId again, so that they can be used like a functional interface
* the `span_` prefix is dropped, that this API deal with OpenTracing span's can be deduced from the module name, anyway
* the `p` prefix is not longer used, while this API might use the process dictionary, it could also use other, most likely configurable backends

**Comment by Licenser** [comment]
I personally prefer the current implementation, it allows more fredom to know to use it. using the f-functions allows to implement something like that without much trouble. However a api like that would not allow the oposite.

That siad I could totally see an extra model like that around using the current API to active that. I did make a own API wrapper for my use too ;).

**Comment by cmullaparthi** [comment]
I like the idea @RoadRunnr. In the 0.2.0 release I've already removed the span_ prefix but I've made the functional API the "main" API and made the process dict API accessible through the  ```otter_span_pdict_api``` module. But I can see how to blend the two. I'll make this part of the 0.3.0 release.

**Comment by cmullaparthi** [comment]
@RoadRunnr @Licenser take a look at the [api_experiment](/Bluehouse-Technology/otter/tree/api_experiment) branch and let me know what you think.

**Comment by RoadRunnr** [comment]
Mostly like the API, two things:

1. Currently the span_id() is defined as integer() and exported. I would prefer if it where a opaque data type. The caller should make no assumptions about it's content. This allows us to change it whenever we see a need for that.

2. Source filtering: There should a way to not start a span when otter:span_start is called, e.g. when we sample only every n-th span. In that case a special span_id should be returned that prevents further processing of span events. otters uses `undefined` for that and has a special match for each function.

    I would further like to have test function that can be used to determine if processing for a span is active or not. Something like:

    ```
    case otter:span_is_active(SpanId) of
        true -> Log = do_something_expensive(),
                 otter:span_log(SpanId, Log),
       ...
    end
    ```

    And if possible a define() in header that can be used in guards:

    ```
    -define(SPAN_IS_ACTIVE(SpanId), is_integer(SpanId)).
    ...
    trace_function(SpanId, State) when ?SPAN_IS_ACTIVE(SpanId) ->
        Log = do_something_expensive(Span),
        otter:span_log(SpanId, Log),
        ok;
    trace_function(_SpanId, _State) ->
        ok.

**Comment by cmullaparthi** [comment]
>  Currently the span_id() is defined as integer() and exported. I would prefer if it where a opaque data type. The caller should make no assumptions about it's content. This allows us to change it whenever we see a need for that.

Yes, this should be easy enough to do, I'll do it.

> Source filtering: There should a way to not start a span when otter:span_start is called, e.g. when we sample only every n-th span. In that case a special span_id should be returned that prevents further processing of span events. otters uses undefined for that and has a special match for each function.

Agree. I'll work on such a feature. Obviously I'll take a pull request as well if someone does it before me :-)

**Comment by fholzhauser** [comment]
We could perhaps add a new span_id API and use ETS to store spans in progress. Would need some extra work to make sure not properly finished spans don't stick around in the table. Probably it would not improve the code intrusion an awful lot (comparing to the functional API), as we still need to pass around the span id. What I like with it though is that it might open a possibility to correlate running spans (perhaps a long shot, I didn't give it a lot of thought). It sounds more like a slightly different (non-functional) version of the functional API. I'm not completely convinced yet. Could someone pls. convince me that it is useful? Regarding the process API we could also add extra functions that take the span name as extra argument. It stretches the conceptual model I had in mind for the process API originally, but apparently it is interesting. I think different kind/name spans make some sense in the process. Having multiple instances (i.e. span ids) of the same span kind/name in a process OTOH sounds dodgy. Regarding sending every n-th span, while I'm aware it is in the opentracing descriptions, I'm not entirely convinced about the usefulness of it. Especially not without having the possibility to convey the corresponding flag to the downstream systems to influence the trace generation in those. Quite some points, I'll open issues for each to discuss them further.

**Comment by Licenser** [comment]
the issue with ETS tables it that it will seriously degrade performance. having a ETS lookup/read and Write for every time a log is added, or tag is set will have a serious impact compared to the functional interface.

**Comment by fholzhauser** [comment]
Thanks for your comment Heinz, of course it has a price (as most functionality). One of the reasons it was considered and not done originally.  At the moment I also don't see practical benefits of this API and I don't think it is worth doing it, but perhaps someone has a really good argument or use case, which is why I opened #13. Let's discuss this particular one further there. As I see the idea came up to work around the single span limitation of the process API which I'm hoping to fix with #14 and let this one go.

**Comment by fholzhauser** [comment]
New API (#13) implemented in 0.3.0
